doc/source/devel/eos-fluidprops.rst

```diff
@@ -0,0 +1,497 @@
+
+  EnzoEOSVariant my_eos_variant;
+
+Then the variable ``my_eos_variant`` holds a default-constructed instance of :cpp:class:`!EnzoEOSVariant`.
+At the time of writing this, this object will contain an instance of an :cpp:class:`!EnzoEOSIsothermal`, but that is an implementation detail that may change over time.
+
+Like :cpp:class:`!std::variant`, :cpp:class:`!EnzoEOSVariant` also has value-like semantics.
+This means that any time you perform a copy on an instance of :cpp:class:`!EnzoEOSVariant`, it's a deepcopy.
+
+.. code-block:: c++
+
+  const EnzoEOSVariant& eos_variant = enzo::fluid_props()->eos_variant();
+
+  // make a copy of eos_variant
+  EnzoEOSVariant my_eos_variant = eos_variant;
+
+  
+Any mutations to the contents of ``my_eos_variant`` will not affect the contents of ``enzo::fluid_props()->eos_variant()``.
+Examples might include:
+
+  * changing the type of object stored within ``my_eos_variant`` (if it initially stores an instance of :cpp:class:`!EnzoEOSIsothermal`, we could replace it with an instance of :cpp:class:`!EnzoEOSIdeal`)
+
+  * mutating the attributes of a stored object within ``my_eos_variant`` (one could imagine mutating the value of gamma stored within a :cpp:class:`!EnzoEOSIdeal` instance)
+
+**As an aside**, the API of :cpp:class:`!EnzoPhysicsFluidProps` is designed so that the user can't accidentally mutate the PE's eos object (you can only mutate copies of that object).
+
+    .. _eos-timestep-example:
+
+Using ``EnzoEOSVariant`` (A concrete example)
+---------------------------------------------
+
+Many hydro methods need to determine the maximum timestep that they allow.
+In the process, they may need to compute:
+
+.. math::
+
+  C_0 \times \min\left(\frac{\Delta x}{|c_{s,ijk}+ v_{x, ijk}|},
+                       \frac{\Delta y}{|c_{s,ijk}+ v_{y, ijk}|},
+                       \frac{\Delta z}{|c_{s,ijk}+ v_{z, ijk}|}\right)
+
+where :math:`C_0` is the courant factor (a constant between 0 and 1) and :math:`\Delta x,\, \Delta y,\, \Delta z` specify cell widths.
+
+The following code snippet shows a somewhat simplified example of how you might do this.
+It assumes that the pressure field has been precaluated it will abort if the program does not use an ideal eos:
+
+.. code-block:: c++
+
+   double timestep(CelloArray<const enzo_float, 3> density,
+                   CelloArray<const enzo_float, 3> velocity_x,
+                   CelloArray<const enzo_float, 3> velocity_y,
+                   CelloArray<const enzo_float, 3> velocity_z,
+                   CelloArray<const enzo_float, 3> pressure,
+                   double dx, double dy, double dz,
+                   double courant_factor)
+   {
+     // the program aborts with an error if Enzo-E was not configured with an
+     // ideal eos (as an aside, we are technically making a copy of the eos
+     // here - that should be totally fine sinces its just a memcpy)
+     const EnzoEOSIdeal eos = enzo::fluid_props()->eos_variant().get<EnzoEOSIdeal>();
```

Comment:
The short answer is: NOT YET. It's absolutely possible and it was originally a major motivation for introducing the ``visit`` function. However, in practice it's not worth doing right now when we only have two equations of state and while we are using C++11.

Here are 2 examples (while they are possible in C++11, I've opted not to show that right now because that is more cumbersome).

If every EOS class has a method with the signature ``enzo_float EOSClass::sound_speed(enzo_float density, enzo_float pressure)``, then we can write easily write the following snippet in C++14 that will do this:

```c++

   double timestep(CelloArray<const enzo_float, 3> density,
                   CelloArray<const enzo_float, 3> velocity_x,
                   CelloArray<const enzo_float, 3> velocity_y,
                   CelloArray<const enzo_float, 3> velocity_z,
                   CelloArray<const enzo_float, 3> pressure,
                   double dx, double dy, double dz,
                   double courant_factor)
   {
     auto lambda_func = [=](auto eos) -> double
     {
       const int mx = density.shape(2);
       const int my = density.shape(1);
       const int mz = density.shape(0);

       double dt = std::numeric_limits<double>::max();
       for (int iz = 0; iz < mz; iz++) {
         for (int iy = 0; iy < my; iy++) {
           for (int ix = 0; ix < mx; ix++) {

             double cs = (double) eos.sound_speed(density(iz,iy,ix),
                                                  pressure(iz,iy,ix));
             double abs_vx = std::fabs((double) velocity_x(iz,iy,ix));
             double abs_vy = std::fabs((double) velocity_y(iz,iy,ix));
             double abs_vz = std::fabs((double) velocity_z(iz,iy,ix));
             double tmp = std::min(std::min(dx/(abs_vx + cs),
                                            dy/(abs_vy + cs)),
                                   dz/(abs_vz + cs));
             dt = std::min(dt, tmp);
           }
         }
       }
       return dt;
     };
     double dt = enzo::fluid_props()->eos_variant().visit(lambda_func);
     return courant_factor * dt;
    }
```

In practice, it doesn't make sense for every EOS to have a method that computes the sound speed from both density and pressure. In the future, one could imagine modifying ``EnzoEOSIsothermal`` so that it has the method ``enzo_float EnzoEOSIsothermal::sound_speed()``. In C++17 we could write a new function that makes use of that new functionallity:

```c++

   double timestep(CelloArray<const enzo_float, 3> density,
                   CelloArray<const enzo_float, 3> velocity_x,
                   CelloArray<const enzo_float, 3> velocity_y,
                   CelloArray<const enzo_float, 3> velocity_z,
                   CelloArray<const enzo_float, 3> pressure,
                   double dx, double dy, double dz,
                   double courant_factor)
   {
     auto lambda_func = [=](auto eos) -> double
     {
       using T = std::decay_t<decltype(arg)>;
       const int mx = density.shape(2);
       const int my = density.shape(1);
       const int mz = density.shape(0);

       double dt = std::numeric_limits<double>::max();
       for (int iz = 0; iz < mz; iz++) {
         for (int iy = 0; iy < my; iy++) {
           for (int ix = 0; ix < mx; ix++) {

             double cs;

             if constexpr (std::is_same_v<T, EnzoEOSIdeal>) {
               cs = (double) eos.sound_speed(density(iz,iy,ix),
                                             pressure(iz,iy,ix));
             } else if constexpr (std::is_same_v<T, EnzoEOSIsothermal>) {
               cs = (double) eos.sound_speed();
             } else {
               static_assert(false, "can't handle specified EOS");
             }

             double 
             double abs_vx = std::fabs((double) velocity_x(iz,iy,ix));
             double abs_vy = std::fabs((double) velocity_y(iz,iy,ix));
             double abs_vz = std::fabs((double) velocity_z(iz,iy,ix));
             double tmp = std::min(std::min(dx/(abs_vx + cs),
                                            dy/(abs_vy + cs)),
                                   dz/(abs_vz + cs));
             dt = std::min(dt, tmp);
           }
         }
       }
       return dt;
     };
     double dt = enzo::fluid_props()->eos_variant().visit(lambda_func);
     return courant_factor * dt;
    }
```

Note that the if-statements in the for-loop are called "constexpr-if-statements". As I understand it, they will not produce branching within the for-loop; instead separate versions of the lambda expression are automatically generated for each possible EOS type (so effectively there is a branch outside of the if-statement).